### PR TITLE
Dedicated SKU column (for issue #495)

### PR DESCRIPTION
**Pull request by AWS-ZachErdman**
**Lead Maintainer**: Zach
### Documents: 
* Folder #495 : [SKU Properties & Providers Alignment](https://drive.google.com/drive/u/0/folders/11CDin-7o7KNGrdCzgq7gdVwQFcJ2l8aq) by Zach moved to folder #495
* Document: [SKU Details from Providers (FOCUS)](https://docs.google.com/document/d/1PdF8VQ5Ad4kDmrTlSEzWGOitwyEeQHR5B13hPWArMg0/edit#heading=h.ezdq0hltg6)
* Spreadsheet: [24.08.08 - Alternative-Column-Names-for-UsageType](https://docs.google.com/spreadsheets/d/1-n_-z1bWKKwEbSXfiICHl0ff-uC9qAdDaDYhQGsagEM/edit?usp=sharing)
* PR [#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516) Initial Draft of UsageType
* Approval date: Sep 05

### CODE REVIEW

**Comment by jpradocueva** [comment]
### Action Items from TF-2 call on Aug 8th:
- [ ] [TF2- [#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Zach to update the PR  [#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)  with a more detailed description and potential names for the new column.
- [ ] [TF2- [#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Tim to gather additional feedback on the prioritization of features. This should address the need for better documentation and decision-making transparency.
- [ ] [TF2- [#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Tim will provide feedback about the new column directly to the pull request
- [ ] [TF2- [#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] A spreadsheet will be created to list potential names and descriptions for the new column, allowing for community voting and feedback.
- [ ] [TF2- [#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)]  The team to revisit the introduction of additional columns in future versions based on user feedback and evolving needs.

**Comment by jpradocueva** [comment]
### Action Items TF-2 call on 21 Aug:
- [ ] [TF2-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516) [Spreadshet](https://docs.google.com/spreadsheets/d/1-n_-z1bWKKwEbSXfiICHl0ff-uC9qAdDaDYhQGsagEM/edit?usp=sharing)] Zach @AWS-ZachErdman : To finalize the document with the proposed column names and structure and share it with the larger group for review.
- [ ] [TF2-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516) [Spreadshet](https://docs.google.com/spreadsheets/d/1-n_-z1bWKKwEbSXfiICHl0ff-uC9qAdDaDYhQGsagEM/edit?usp=sharing)]  All Members: To review the document and provide feedback on the proposed naming conventions by the next meeting.

**Comment by jpradocueva** [comment]
### Action Items, Members' call on Aug 22:
[Members-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Zach @AWS-ZachErdman : To prepare a poll and suggested names for the next meeting.

**Comment by jpradocueva** [comment]
### Action Items from Maintainers' call on Aug 26th: 
- [ ] [Maintainers-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516) ] Zach - @AWS-ZachErdman : To update the PR with the final name and revised language.

File: specification/columns/usagetype.md
Comment (ijurica) [inline specification/columns/usagetype.md:3]: I believe this column should focus on the SKU (an underlying tracked and measured component of a resource or service) rather than the service.

I'm unsure whether it should provide only the core function or aspect of the SKU or include additional details as well. If we plan to introduce separate columns for SKU Size, SKU Region (potentially in a JSON format), etc., I would suggest focusing on the core function of the SKU for this column. 

That said, I understand that providers might initially map native columns that include more than just the core function, but I hope that over time this additional information will be parsed out and provided in separate columns. 
(Some practitioners and vendors already parse out this information, and it would be great if providers could do the same for us :) )

Potential mapping candidates:
- AWS: UsageType or [UsageTypeGroup](https://docs.aws.amazon.com/cost-management/latest/userguide/ce-filtering.html)
- Microsoft: Meter or MeterCategory?

**Comment by jpradocueva** [comment]
### Action Items from TF-2 call on Aug 28:
- [ ] [TF2-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Zach - @AWS-ZachErdman : Update the PR to reflect the new column name "SKU Meter" and refine the description to align with the discussion.
- [ ] [TF2-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] All Members: Review the updated PR and vote on the final column name and description by the next meeting.
- [ ] [TF2-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Zach - @AWS-ZachErdman : Reach out to Sean, Irina, Michael, Tim, and other relevant members to gather their approval on the updated PR before the next meeting.

**Comment by udam-f2** [comment]
@AWS-ZachErdman Seems like we'll need a good amount of supporting content to be added here to capture all the discussions and how we ended up getting around to this.

**Comment by jpradocueva** [comment]
### Action Items from Maintainers' call on Sep 2:
- [ ] [Maintainers-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Team members to gather further feedback on the preferred terminology and be prepared to discuss it in the next meeting.

File: specification/columns/skumeter.md
Comment (ijurica) [inline specification/columns/skumeter.md:15]:
```suggestion
SKU Meter
```

File: specification/columns/skumeter.md
Comment (ijurica) [inline specification/columns/skumeter.md:7]: I suggest we tighten up the requirements and align the nullability with SkuId (I believe all Providers can describe the capability of the component they are measuring and charging). 

Should we opt for that, even though for some other columns we avoided referencing the SkuId and/or SkuPriceId columns, I believe that, just like we referenced SkuPriceId in the case of SkuPriceDetails, it is okay to reference SkuId in the case of SkuMeter.

Adding quote from SkuPriceDetails for reference:

> The SkuPriceDetails column MUST be present in the billing data when the provider includes a SkuPriceId.

```suggestion
The SkuMeter column adheres to the following requirements:

 * SkuMeter MUST be present in the billing data when when the provider includes a [SkuId](#skuid).
 * SkuMeter MUST be of type String.
 * SkuMeter MUST be null when SkuId is null and MUST NOT be null when SkuId is not null.
```

Comment (AWS-ZachErdman) [reply specification/columns/skumeter.md:7]: For AWS, not every SKU will have a MeterName (i.e. usage type). I'd be happy to say this should be null when SkuId is null, but I wouldn't want to force a value when SkuId isn't null.

File: specification/columns/skumeter.md
Comment (AWS-ZachErdman) [reply specification/columns/skumeter.md:15]: fixed and updated to SKU Meter Name, thanks!

**Comment by AWS-ZachErdman** [comment]
> @AWS-ZachErdman Seems like we'll need a good amount of supporting content to be added here to capture all the discussions and how we ended up getting around to this.

Added the content for review.

File: specification/columns/usagetype.md
Comment (ijurica) [reply specification/columns/usagetype.md]: Since the column definition has changed, this comment is outdated, and I'm resolving it.

**Review by ijurica — APPROVED** [review]

File: specification/columns/skumetername.md
Comment (cnharris10) [inline specification/columns/skumetername.md:14]: I've never seen an examples section.  Are we adding this pattern?  Otherwise, I've seen examples as part of the top-level description

File: supporting_content/columns/skumetername.md
Comment (cnharris10) [inline supporting_content/columns/skumetername.md:1]:
```suggestion
# Column: SKUMeterName
```

**Comment by jpradocueva** [comment]
### Action Items from TF-2 call on Sep 4:
- [ ] [TF2-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Zach, @AWS-ZachErdman : Update the PR with "SKU meter" as the finalized term and make the necessary edits for nullability conditions.

File: specification/columns/skumetername.md
Comment (cnharris10) [inline specification/columns/skumetername.md:10]: Upfront, I realize that the answer here is probably, "we'll solve this in iteration", and this is not meant to block this pr.

I don't love how we're going to be combining coded strings (i.e. AWS) with clause-based strings (i.e. Azure, GCP).

Example
```
AWS: EU1-BoxUsage:m5.2xlarge
Azure: Native Metric Queries API Calls
GCP: N2D AMD Instance Core running in Frankfurt
```

_Ideally_, I'd push for a SkuMeterCode, SkuMeterName pairing where, with the examples above, AWS would populate `SkuMeterCode` and SHOULD provide a clause-based `SkuMeterName`, but Azure/GCP would populate `SkuMeterName` and SHOULD provide a code-based `SkuMeterCode`.

Working with data extensively across these 3 providers, I find classifying by AWS' usage types to be easily filterable because they usually follow a pattern, but I find them hard to decipher.  The opposite is true for Azure/GCP.  They are not easy to filter but easy to understand.

File: supporting_content/columns/skumetername.md
Comment (cnharris10) [inline supporting_content/columns/skumetername.md:10]: Wouldn't this be `sku.description`?

```suggestion
| GCP       | Big Query Billing Export | sku.description                |
```

File: supporting_content/columns/skumetername.md
Comment (ahullah) [inline supporting_content/columns/skumetername.md:42]: +1 on this, don't block for this release but this would be much more directly usable by practitioners

File: specification/columns/skumetername.md
Comment (ahullah) [reply specification/columns/skumetername.md:10]: would this not be better represented by the 'longer term solution' described in the supporting content?

**Review by ahullah — APPROVED** [review]

File: supporting_content/columns/skumetername.md
Comment (flanakin) [inline supporting_content/columns/skumetername.md:34]: Would it make sense to add a naming discussion to this as well? I'd hate for people to look at this and assume Microsoft pushed this name. I for one have concerns with the name based on Microsoft's historical usage of it. I'm not trying to push back, but it would be nice to have some overview of the names we discussed and why we landed on this one.

**Review by flanakin — APPROVED** [review]

File: supporting_content/columns/skumetername.md
Comment (AWS-ZachErdman) [reply supporting_content/columns/skumetername.md:34]: Yeah let's put together another vote for the name with some of the ideas we discussed on slack

**Comment by jpradocueva** [comment]
### Action Items Maintainers' call on Sep 6:
- [ ] [Maintainers-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Zach, @AWS-ZachErdman : Coordinate with Task Force 2 to review and finalize the UsageType draft before the next meeting.
- [ ] [Maintainers-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Reporting Team: Assess how the UsageType changes will impact reporting across different providers.

File: specification/columns/skumetername.md
Comment (ijurica) [reply specification/columns/skumetername.md:9]: - I prefer **precision** in normative requirements, i.e. referencing a column ID whenever possible (similar to guidelines for developers implementing validators). 

- Considering we've specified that _SkuId MUST equal SkuPriceId when a provider does not support an overarching SKU ID construct_, I suggested tying it to SkuId. If the group prefers SkuPriceId, I have no objections. That said, when it comes to the column name and prefix, I still prefer 'Sku' over 'SkuPrice.' :)

File: specification/columns/skumetername.md
Comment (ijurica) [inline specification/columns/skumetername.md:7]: In SkuPriceDetails, we included the following requirement:

> Individual properties (key-value pairs) SHOULD NOT be modified for a given SkuPriceId and SHOULD remain consistent over time.

Would it be helpful to include something similar here as well? For instance:

> SkuMeterName SHOULD NOT be modified for a given SkuId and SHOULD remain consistent over time.

File: supporting_content/columns/skumetername.md
Comment (AWS-ZachErdman) [reply supporting_content/columns/skumetername.md:34]: TF-2 voted for SkuMeter today (9/11/24)

File: specification/columns/skumetername.md
Comment (AWS-ZachErdman) [reply specification/columns/skumetername.md:14]: We have an examples section in SkuPriceDetails!

I'll leave it in for now and we can remove later if needed.

File: specification/columns/skumetername.md
Comment (AWS-ZachErdman) [reply specification/columns/skumetername.md:9]: I agree with Irena, the Column ID enables better conformance evaluation. Will close this for now.

File: supporting_content/columns/skumetername.md
Comment (AWS-ZachErdman) [reply supporting_content/columns/skumetername.md:42]: Got it!

File: specification/columns/skumetername.md
Comment (AWS-ZachErdman) [reply specification/columns/skumetername.md]: Makes sense - added.

**Comment by jpradocueva** [comment]
### Action Items from TF-2 call on Sep 11:
- [ ] [TF1-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Zach, @AWS-ZachErdman : Prepare the PR for the naming revote and include a few additional options for clarity.
- [ ] [TF1-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Michael, @flanakin : Provide alternative names for comparison to the team before the next meeting.

File: specification/columns/skumeter.md
Comment (kk09v) [inline specification/columns/skumeter.md:13]: References "SkuMeterName" vs "SkuMeter" though I know that the naming is up for debate currently

File: specification/columns/skumeter.md
Comment (kk09v) [inline specification/columns/skumeter.md:5]:
```suggestion
Providers often have billing models in which multiple SKUs exist for a given service to describe and bill for different functionalities for that service. For example, an object storage service may have separate SKUs for functionalities such as object storage, API requests, data transfer, encryption, and object management. This field helps practitioners understand which functionalities are being metered by the different SKUs that appear in a FOCUS dataset.
```

Adding "and" to the list.

**Comment by jpradocueva** [comment]
There is a problem compiling the PDF due to missing files: `capacityreservationid.md` and `capacityreservationstatus.md`

image

File: specification/columns/skumeter.md
Comment (AWS-ZachErdman) [reply specification/columns/skumeter.md]: thanks - fixed

**Comment by jpradocueva** [comment]
### Action Items from Members' call on Feb 12:
- [ ] [TF2-[#516](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/516)] Zach, @AWS-ZachErdman : Finalize any outstanding reviews and prepare for next week’s approval.

**Review by flanakin — APPROVED** [review]

**Comment by jpradocueva** [comment]
Approved by the Members' group on the Sep 19 call.

**Review by udam-f2 — APPROVED** [review]

**Pull request merged**
